### PR TITLE
Two Layer Data Model Prototype

### PR DESCRIPTION
**Pull request by roseeichelmann**
Instructions to get the database started and run through the test cases in the `README.md` :sparkles: 

Happy copy :spaghetti: ing!!

Also -- let me know if you need me to wormhole you a copy of the `atd_txdot_locations` table csv.

### CODE REVIEW

File: build_tables.sql [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            WHERE (geometry && ST_SetSRID(ST_MakePoint(NEW.longitude, NEW.latitude), 4326))
+            AND ST_Contains(geometry, ST_SetSRID(ST_MakePoint(NEW.longitude, NEW.latitude), 4326))
+            LIMIT 1 --TODO: This should be temporary until we get our polygons in order
+        );
+        RETURN NEW;
+    END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER update_crash_location_id
+    BEFORE INSERT OR UPDATE ON public.crashes FOR EACH ROW EXECUTE FUNCTION public.update_crash_location_id();
```
Comment (roseeichelmann) [inline build_tables.sql]: this could be made more efficient by actually checking for change to the lat/lon columns on a separate update trigger, but i didnt get around to that heh.

File: README.md [hunk 2]
```diff
@@ -1,11 +1,19 @@
-# VZ Model Workspace
+# VZ Two Layer Model Prototype
+
+![Diagram of the two layer prototype.](diagram.png)
+
+- CRIS users will insert and update into the tables that are prefixed with `cris_`
+- Inserts to the `cris_` tables will trigger copying the record into their respective `crashes` or `units` table
+- VZ users will make edits to the `crashes` and `units` tables
+- The final "source of truth" are the `crashes` and `units` tables
+- Columns that are prefixed with `vz_` were created by VZ and are not columns we receive from CRIS
```
Comment (johnclary) [inline README.md:9]: > Columns that are prefixed with `vz_` were created by VZ and are not columns we receive from CRIS

i appreciate you flagging this shortcoming in our current DB. i agree—let's make it super clear when a column is an official CRIS column vs our own add-on/customization.

File: README.md [hunk 3]
```diff
@@ -25,12 +33,192 @@ docker compose up -d adminer
...
+  COALESCE(crashes.crash_count, 0) AS crash_count,
+  COALESCE(crashes.unit_count) AS unit_count
+   FROM locations
+     LEFT JOIN crashes ON locations.location_id::text = crashes.vz_location_id::text
+  WHERE locations.location_group = 1;
+
+SELECT *
+FROM "locations_with_crash_unit_counts"
+ORDER BY "crash_count" DESC
+LIMIT 100
```
Comment (johnclary) [inline README.md:209]: this works great, and I also tested without the unit count (because that's what I tested in Mike's and my own data model):

```sql
WITH crashes AS (
  SELECT crashes.vz_location_id,
      count(DISTINCT crashes.crash_id) AS crash_count
      FROM crashes
    WHERE true AND crashes.vz_location_id IS NOT NULL
    GROUP BY crashes.vz_location_id
)
  SELECT locations.description,
  locations.location_id,
  COALESCE(crashes.crash_count, 0) AS crash_count
   FROM locations
     LEFT JOIN crashes ON locations.location_id::text = crashes.vz_location_id::text
  limit 100;
```

this is clocking in at 800ms for me 🚀

File: conflict_res.png
Comment (johnclary) [inline conflict_res.png:1]: looks good to me!

File: README.md [hunk 4]
```diff
@@ -25,12 +33,192 @@ docker compose up -d adminer
...
+```sql
+create or replace procedure cris_update_unit (id integer, unit_type integer) language plpgsql as
+$$
+begin
+    raise notice 'CRIS user updating unit type of unit_id % to %', id, unit_type;
+    UPDATE public.cris_units set unit_type_id = unit_type WHERE unit_id = id;
+end;
+$$;
+
+CALL cris_update_unit(2, 1);
```
Comment (johnclary) [inline README.md:130]: This update works fine, but I'm not seeing the results reflected in the public units table:

```sql
select * from public.units where unit_id = 2;

 unit_id | crash_id | unit_type_id
---------+----------+--------------
       2 |        1 |            6
```

I am realizing this requirement is not super clear in our document, but it is in your diagram: if the current version of the `public.units` matches the `public.cris_units` value, then I would expect an update to `cris_units` to propagate to `units`.

As you've implemented it, the `public.units` values can only be changed by VZ users who edit the row directly. CRIS updates never propagate after the initial record insert.

File: README.md [hunk 5]
```diff
@@ -25,12 +33,192 @@ docker compose up -d adminer
...
+    SELECT INTO old_unit_type_id unit_type_id from public.units WHERE unit_id = u_id;
+    SELECT INTO old_unit_type_desc description from public.unit_type_lkp WHERE id = old_unit_type_id;
+    raise notice 'VZ user adding new unit type % with id % to unit_type_lkp table', type_desc, unit_type;
+    INSERT INTO public.unit_type_lkp(id, description) values (unit_type, type_desc);
+    UPDATE public.units set unit_type_id = unit_type WHERE unit_id = u_id;
+    raise notice 'VZ user updated unit type of unit id % from % to %', u_id, old_unit_type_desc, type_desc;
+end;
+$$;
+
+CALL vz_custom_lkp(2, 11111, 'scooter');
```
Comment (johnclary) [inline README.md:152]: nice—i like that you included the steps to backfill this value for existing unit records.

**Review by johnclary — APPROVED** [review]
Rose, this is great! I was able to get all the data and imported once I switch to using Frank's ARM-friendly postgis image.

My initial impression of your data model is that it's nice and easy to follow. The `build_tables.sql` is totally digestible. The tests steps work great, with the one shortcoming that CRIS updates are not flowing into the crashes or units tables. I think that this could be handled with triggers pretty easily.

On the performance side, things are looking very excellent. The heavy-duty queries ran very quickly without issue. This model definitely feels optimized for fast queries :sunglasses:

The simplicity of this "two-layer" model is really something to admire. The only obvious downside I'm seeing is that in order to identify VZ user edits, you have to diff against CRIS and that diff implies VZ user edits. Whereas I believe all of the other data models we review are going to hold a copy of VZ user edits elsewhere. I think we would definitely want to pair this with a change log feature that logged 100% of user edits, but I think we should do that for any data model we use anyway.

TLDR, I think your model is really compelling in it's simplicity—and it would be nice to flesh out the missing functionality for propagating CRIS updates.

File: build_tables.sql [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    location_group smallint DEFAULT 0
+);
+
+CREATE INDEX locations_geometry_index ON public.locations USING gist (geometry);
+
+-- TRIGGERS
+
+-- Trigger on insert to cris table
+
+-- order of these triggers will depend on how cris data is inserted into the db - are crashes first or units?
```
Comment (mddilley) [inline build_tables.sql]: glad that you were thinking about this! I got lost in thinking about it for a bit and landed on crash first and then related records.

File: build_tables.sql [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+CREATE OR REPLACE TRIGGER update_crash_location_id_on_update
+BEFORE UPDATE ON public.crashes
+FOR EACH ROW
+WHEN (OLD.latitude IS DISTINCT FROM NEW.latitude
+OR OLD.longitude IS DISTINCT FROM NEW.longitude)
+EXECUTE FUNCTION public.update_crash_location_id();
+
+-- Trigger for vz_unique_unit_types
+
+CREATE OR REPLACE FUNCTION public.update_unique_unit_types()
```
Comment (mddilley) [inline build_tables.sql:278]: very cool to see that this didn't end up being a bottleneck to seeding in your model! ✨ I struggled here!

File: README.md [hunk 8]
```diff
@@ -25,12 +33,192 @@ docker compose up -d adminer
...
+create or replace procedure vz_update_latlon (id integer, lat double precision, long double precision) language plpgsql as
+$$
+declare old_location text;
+declare new_location text;
+begin
+    raise notice 'VZ user updating lat/long of crash_id %', id;
+    SELECT INTO old_location vz_location_id FROM public.crashes WHERE crash_id = id;
+    UPDATE public.crashes SET latitude = lat, longitude = long WHERE crash_id = id;
+    SELECT INTO new_location vz_location_id FROM public.crashes WHERE crash_id = id;
+    raise notice 'Location updated from % to %', old_location, new_location;
```
Comment (mddilley) [inline README.md:78]: 🙏 this logging in the test cases was so useful!! makes me wanna go back and update mine. 😄

File: README.md [hunk 9]
```diff
@@ -25,12 +33,192 @@ docker compose up -d adminer
...
+
+SELECT *
+FROM "locations_with_crash_unit_counts"
+ORDER BY "crash_count" DESC
+LIMIT 100
+```
+
+11. Create a test case, or simply write out the steps that would be involved if we wanted to add a new editable column to crashes
+
+We would prefix the column with `vz_` and add it only to the `public.crashes` table
```
Comment (mddilley) [inline README.md:214]: i'm realizing that we didn't all test a column addition the same way. there are different ways a new editable column is added which include:
- VZ team wants to make an existing CRIS column editable
- VZ team wants to make a custom editable field 

I didn't check out that second one so I'm glad that you explored it. 🚀

**Review by mddilley — COMMENTED** [review]
I love that this model takes the least amount of code to get going. 💪  Also, not having views for the main tables gives it a performance advantage. I also see a short path from where we are now to a model like this.

Tradeoffs that I see (there is overlap with john's model so excuse the copypasta!):
- like john mentioned - cris updates on fields that are not-yet-edited will be hidden by the stale values in the `crashes` or `units` tables. BUT i think that this is accounted for in your conflict resolution diagram if I'm looking at it right.
- sorting out what is an edit and what is an original cris value could take more work, but we could also be underestimating how much a well-designed change log will help us with this!
- denormalization (but in an intentional and documented way) - i think the key is that if we understand how the denormalization works and it provides a performance benefit then it is a-okay 👍
- custom lookup values in the same table as cris lookup values. This is exactly what I did too, and I think the same thing that I said about denormalization. If we know why and how it works and it keeps things simple, I am all for it!

benefits:
- native table performance perk 🏎️💨 All of your queries were speedy for me!
- two layers to compare VZ and CRIS values
- implementation. i see a path to this model by building off our current DB

File: README.md (hunk 4 above)
Comment (roseeichelmann) [reply README.md:130]: Yeahhh i definitely forgot to write this trigger but it should be there darn :face_with_spiral_eyes: i feel like i need to go back and add this for my model to be complete

**Comment by johnclary** [comment]
I had one other thought about Rose's model vs all the others. If we ever want/need to be able to edit multiple field values through the VZE, it is going to be a lot easier to write mutations under Rose's model. For example, take this record:

|column|value|
|-|-|
| crash_id| 1|
| primary_address| 123 Fake St |
| secondary_address | 200 Main St |
| road_type_id| 2|

If a user is editing a form that includes `primary_address` and `secondary_address`, and the user edits only the `secondary_address`, we will need to write extra editor code so that the form submission only includes  the modified `secondary_address`, but not the `primary_address`, in the VZ record mutation. Otherwise, our VZ edits rows would accumulate data that has not actually been edited. This is not a concern under Rose's model because all edits mix with the existing CRIS values in the second layer.

We don't have the need to edit multiple fields right now, but it's definitely a need that has crossed my mind as we look at UI enhancements for editing certain crash fields.

File: build_tables.sql [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+);
+
+create index on public.units (crash_id);
+
+-- LOCATIONS
+CREATE TABLE public.locations (
+    location_id character varying PRIMARY KEY,
+    description text NOT NULL,
+    address text,
+    last_update date DEFAULT now() NOT NULL,
```
Comment (frankhereford) [inline build_tables.sql:67]: The production locations table has a `metadata` field in between `address` and `last_update` - I'm totally game to remove it, but I was curious if it was problematic in some way? I added it back in to make loading up this table easier, and I'm curious if that's OK.

**Review by frankhereford — APPROVED** [review]
Rose, this is great. I strongly second everyone's observations about how you've kept things in normal tables and out of views, and that really shows in terms of performance. :rocket: 

I also appreciate how you've taken a novel approach to defining procedures to interact with the data. Nothing's going to be more customization than that, and even though the prototype requirements only ask us to do a basic `coalesce()`, in reality, there are going to be more complex logic around how certain columns are computed based on what VZ has offered, what CRIS has offered, and what other columns state exists that the computed value may depend on. Your leaning into the procedures leaves a really clean way to execute that logic.

Here are some notes I took while I went through your steps. Thanks again for this work!

# DB loading

* In the locations table, the metadata column appears to be removed, is it problematic?

# Test Steps

## Changing a crash location as VZ

* Do you envision creating a procedure to change any column's value?
* This is way out of the scope of the prototype, but I wonder if these procedures be called via the normal graphql-engine endpoint, or if we'd want to switch to the SQL endpoint.

## Custom lookup values

* I wonder what you think about ways to not encode meaning into the ID value, like `> 11111` means VZ generated values. 

## All crash query

```sql
with all_crashes as (
SELECT * FROM public.crashes)
select * from all_crashes order by random() limit 10;
```

That runs in ~ 4 seconds on the test machine, and that's quick!

## All locations

```sql
with all_locations as (
SELECT *
FROM "locations_with_crash_unit_counts"
ORDER BY "crash_count" DESC
) select * from all_locations order by random() limit 10;
```

~ 10 seconds, which I think is more a reflection of the IO speed of the machine and not the DB schema. It's fast.

## Adding a custom VZ column

* This is really slick; it just couldn't be simpler.

File: README.md [hunk 11]
```diff
@@ -25,12 +33,226 @@ docker compose up -d adminer
...
+
+```sql
+ALTER TABLE public.crashes ADD COLUMN vz_law_enforcement_num integer;
+UPDATE public.crashes SET vz_law_enforcement_num = 1 WHERE crash_id = 1;
+SELECT * FROM public.crashes WHERE crash_id = 1;
+```
+
+12. [Optional] Create a test case or simply describe the mechanism to support a conflict management system as described in the functional requirements.
+
+Here are some commands to just test the functionality of CRIS updating a value that doesn't yet have any VZ edits, and then updating a value that does have VZ edits.
```
Comment (roseeichelmann) [inline README.md:224]: for testing updates from cris!

File: build_tables.sql [hunk 12]
```diff
@@ -182,6 +181,66 @@ CREATE OR REPLACE TRIGGER copy_crash_on_cris_insert
...
+declare
+    new_jb jsonb := to_jsonb(NEW); 
+    old_jb jsonb := to_jsonb(OLD); 
+    vz_table_jb jsonb;
+    running_column text;
+begin
+    SELECT row_to_json(crashes) INTO vz_table_jb FROM crashes WHERE crash_id = NEW.crash_id; 
+    for running_column in select jsonb_object_keys(new_jb) loop
+        if (new_jb -> running_column <> old_jb -> running_column) then
+            if (old_jb -> running_column = vz_table_jb -> running_column) then
```
Comment (roseeichelmann) [inline build_tables.sql:200]: if the current vz table value equals the old cris table value, we assume there has been no vz edit, and carry over the update from CRIS into the table containing vz edits

File: build_tables.sql [hunk 13]
```diff
@@ -182,6 +181,66 @@ CREATE OR REPLACE TRIGGER copy_crash_on_cris_insert
...
+end;
+$function$
+;
+
+CREATE OR REPLACE TRIGGER on_cris_crashes_update
+    BEFORE UPDATE ON cris_crashes
+    FOR EACH ROW
+    EXECUTE FUNCTION public.on_cris_crashes_update();
+
+CREATE OR REPLACE FUNCTION public.on_cris_units_update()
```
Comment (roseeichelmann) [inline build_tables.sql:215]: fwiw these could be combined into one trigger function and use the `TG_TABLE_NAME` to figure out which table to be updating and stuff buuuuut i didnt do that here, possible enhancement tho if we were to actually pursure this method so we only have one of these kind of trigger functions to keep track of for all the tables

File: build_tables.sql [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+    for running_column in select jsonb_object_keys(new_jb) loop
+        if (new_jb -> running_column <> old_jb -> running_column) then
+            if (old_jb -> running_column = vz_table_jb -> running_column) then
+                EXECUTE format('UPDATE public.crashes SET %I = $1.%I WHERE crash_id = $1.crash_id;', running_column, running_column) USING  NEW;
+            end if;
+        end if;
+    end loop;
+    return NEW;
+end;
+$function$
```
Comment (johnclary) [inline build_tables.sql:207]: this is really, really slick! i love it. 

one thing i think we'd want to do is plan for handling multiple column edits in one update statement.  i was able to do this by adapting yours to fit my own data structure, and tweaked it to store the updates in an array and join them all together at the end. you'll get the idea:

```sql
create or replace function db.crashes_cris_update()
returns trigger
language plpgsql
as $$
declare
    new_jb jsonb := to_jsonb (new);
    old_jb jsonb := to_jsonb (old);
    vz_record_jb jsonb;
    running_column text;
    updates_todo text [] := '{}';
    update_stmt text := 'update db.crashes ';
begin
    -- get the VZ record as jsonb
   SELECT row_to_json(crashes_vz) INTO vz_record_jb FROM db.crashes_vz WHERE db.crashes_vz.crash_id = NEW.crash_id; 

    -- for every key in the new json object
    for running_column in select jsonb_object_keys(new_jb) loop
        -- if the new value doesn't match the old
        if(new_jb -> running_column <> old_jb -> running_column) then
            -- see if the vz record has a value for this field
            if (vz_record_jb ->> running_column is  null) then
                -- this value is not overridden by VZ
                -- so we need to append this column/value to our todos
                updates_todo := updates_todo || format('set %I = $1.%I', running_column, running_column);
            end if;
        end if;
    end loop;
    -- if we have updates todo
    if(array_length(updates_todo, 1) > 0) then
        -- complete the update statement by joining all `set` clauses together
        update_stmt := update_stmt
            || array_to_string(updates_todo, ',')
            || format(' where db.crashes.crash_id = %s', new.crash_id);
        raise notice 'Executing statement: %', update_stmt;
        -- execute `using new` ensures that all values are appropriately quoted based on their type
        execute (update_stmt) using new;
    else
        raise notice 'No changes to truth record detected.';
    end if;
    return null;
end;
$$;
```

Comment (roseeichelmann) [reply build_tables.sql:207]: @johnclary woahhh this is super frickin cool thanks for sharing, love that ur thinking ahead about multiple column changes at once :sparkles:

Comment (mddilley) [reply build_tables.sql:207]: +1 - so excited to see the jsonb approach along with plucking and comparing the relevant values that you came up with here, Rose! ❤️ Also, John adding support for multiple column updates is 💯. It is a use case that we should prepare for!
